### PR TITLE
Nk/driver temperature shutdown

### PR DESCRIPTION
**Pull request by nkrackow**
This PR adds a temperature monitoring task to check the Driver and headboard temperatures and turn off the channels to prevent runaway heating and damage in Driver. https://github.com/sinara-hw/Driver/issues/18

It was thereby necessary to refactor the way the interlock works by adding the feature that tripping the interlock implicitly also turns off both channels (aka triggers the state machines to transition through to the `Disabled` state). This means that the miniconf `output_enabled` state for both channels can be out of sync with the actual output state machine. This is a necessary sacrifice to be able to turn off the outputs and I think it doesn't matter that much. 
You still have to first set the miniconf setting to disabled first before resetting the laser interlock. So nothing changed from a users perspective except for no self destruction during a high output current interlock event. 
implements https://github.com/quartiq/stabilizer/issues/706

### CODE REVIEW

File: src/bin/driver.rs [hunk 1]
```diff
@@ -770,12 +822,26 @@ mod app {
         {
             if (read > interlock_current) & interlock_asserted & output_en {
                 let ch = Channel::try_from(i).unwrap();
-                c.shared.laser_interlock.lock(|ilock| {
-                    ilock.set(Some(Reason::Overcurrent(Condition {
-                        channel: ch,
-                        threshold: *interlock_current,
-                        read: *read,
-                    })))
+                (c.shared.laser_interlock).lock(|ilock| {
```
Comment (nkrackow) [inline src/bin/driver.rs:825]: The nested locking is done due to https://github.com/rtic-rs/rtic/issues/713

**Comment by nkrackow** [comment]
The hardcoded valid temperature range obviously has to be discussed and potentially adapted still.

File: src/hardware/driver/output.rs [hunk 2]
```diff
@@ -310,14 +310,20 @@ where
...
+            sm::States::RampCurrent => {
+                (Some(Output::<I2C>::RAMP_DELAY), result)
+            }
+            sm::States::SelftestShunt => {
+                (Some(Output::<I2C>::SET_DELAY), result)
+            }
+            sm::States::SelftestShort => {
+                (Some(Output::<I2C>::SET_DELAY), result)
+            }
+            _ => (None, result),
```
Comment (jordens) [inline src/hardware/driver/output.rs:326]: `(match .. {}, result)`

File: src/hardware/driver/mod.rs [hunk 3]
```diff
@@ -123,11 +143,27 @@ impl LaserInterlock {
...
+        reason: Option<Reason>,
+        output_state: &mut [output::sm::StateMachine<output::Output<I2c1Proxy>>;
+                 2],
+    ) -> [Option<fugit::MillisDuration<u64>>; 2] {
         // only update if no reason yet or if clearing (remember first reason)
         if self.reason.is_none() || reason.is_none() {
             self.pin.set_state(reason.is_none().into());
+            // disable both outputs if interlock is tripped
             self.reason = reason;
+            if self.reason.is_some() {
```
Comment (jordens) [inline src/hardware/driver/mod.rs:157]: A bit hard to parse mentally. Maybe clearer with a `match` or an inversion of thw two `if` s?

File: src/bin/driver.rs [hunk 4]
```diff
@@ -719,13 +758,24 @@ mod app {
...
+                let delays = ilock.set(Some(reason), state);
+                delays.iter().enumerate().for_each(|(i, delay)| {
+                    if let Some(delay) = delay {
+                        handle_output_tick::spawn_after(
+                            delay.convert(),
+                            Channel::try_from(i).unwrap(),
+                        )
+                        .unwrap();
+                    }
+                });
```
Comment (jordens) [inline src/bin/driver.rs:776]: This looks like something refactorable (it shows up five times). Or at least the part except the `spawn_after`.

**Comment by nkrackow** [comment]
Not necessary any more due to https://github.com/quartiq/board_modifications/issues/6#issuecomment-1486856599 and https://github.com/quartiq/stabilizer/commit/136f0f54deb44afd4b8de392f94e0663c7facccf
